### ISSUE TITLE
WCF StreamBodyWriter 4.0-4.5 retargeting change

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [dotnet/docs#13940]
Please add the following to the list of migration issues.

If an implementation of `System.ServiceModel.Channels.BodyWriter.OnWriteBodyContents(XmlDictionaryWriter writer)` intends to output a stream, it must call `writer.WriteStartElement("Binary")`, `writer.WriteBase64(buffer, offset, count)`, and perhaps eventually `writer.WriteEndElement`. So the message body is conceptually XML `<Binary>…base64…</Binary>`, although WCF normally manages to skip the Base64 encoding and decoding.

The `System.ServiceModel.Channels.StreamBodyWriter` class handles this conceptual Base64 encoding. Applications can define a derived class that overrides `StreamBodyWriter.OnWriteBodyContents(Stream stream)` and writes the data to the stream. In .NET Framework 4.0 however, `StreamBodyWriter` does not call `writer.WriteStartElement("Binary")` even though it is required by `System.ServiceModel.Channels.XmlByteStreamWriter`. The derived class thus has to override `StreamBodyWriter.OnWriteBodyContents(XmlDictionaryWriter writer)`, like so:

```C#
protected override void OnWriteBodyContents(System.Xml.XmlDictionaryWriter writer)
{
    writer.WriteStartElement("Binary");
    base.OnWriteBodyContents(writer);
    writer.WriteEndElement();
}
```

If such an application targeting .NET Framework 4.0 is then run on .NET Framework 4.5 or later, `StreamBodyWriter` activates a quirk so that the application still works. However, if the application is retargeted to .NET Framework 4.5 or later, WCF discards the data written to the stream. No exception is thrown. This happens because `System.ServiceModel.Channels.StreamBodyWriter.XmlWriterBackedStream.Write(byte[] buffer, int offset, int count)` sees that `writer.WriteState` is `WriteState.Element`, rather than `WriteState.Content` or `WriteState.Start`.

Thus, when the application is being retargeted to .NET Framework 4.5 or later, the override of `OnWriteBodyContents(XmlDictionaryWriter)` should be removed. Alternatively, I think the following code will work regardless of whether the application targets .NET Framework 4.0, 4.5, or a later version.

```C#
protected override void OnWriteBodyContents(System.Xml.XmlDictionaryWriter writer)
{
    writer.WriteStartElement("Binary");
    writer.WriteString(""); // Changes writer.WriteState from WriteState.Element to WriteState.Content.
    base.OnWriteBodyContents(writer);
    writer.WriteEndElement();
}
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 8186b5d8-b712-73a1-cabb-fff3d92945c7
* Version Independent ID: cc3546b7-79c1-b37d-38c1-56eb35fb79d5
* Content: [Retargeting Changes for Migration from .NET Framework 4.0 to 4.5](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/retargeting/4.0-4.5#feedback)
* Content Source: [docs/framework/migration-guide/retargeting/4.0-4.5.md](https://github.com/dotnet/docs/blob/master/docs/framework/migration-guide/retargeting/4.0-4.5.md)
* Product: **dotnet-framework**
* Technology: **dotnet-appcompat**
* GitHub Login: @chlowell
* Microsoft Alias: **ronpet**

---
[Associated WorkItem - 199399](https://dev.azure.com/msft-skilling/Content/_workitems/edit/199399)

**Comment by KalleOlaviNiemitalo** [comment]
A related Stack Overflow question: [Breaking change in class XmlWriterBackedStream with KB2901983](https://stackoverflow.com/questions/28068944/breaking-change-in-class-xmlwriterbackedstream-with-kb2901983)

It refers to [KB2901983: The Microsoft .NET Framework 4.5.2 for Windows and the corresponding language packs are available](https://support.microsoft.com/en-us/help/2901983/the-microsoft-net-framework-4-5-2-for-windows-and-the-corresponding-la).

So, perhaps the behaviour was changed in .NET Framework 4.5.2. The target version threshold for the compatibility quirk is 4.5 rather than 4.5.2, however.

**Comment by rpetrusha** [comment]
@mconnew, can you confirm that this should be added to the compatibility documentation?

**Comment by mconnew** [comment]
The behavior change is correct. I'll need to spend some time validating the exact details of the code change required in customer code, but from memory I believe it's correct. I'll update next week whether it's good as is or if any changes are needed.

**Comment by KalleOlaviNiemitalo** [comment]
> When deriving from the class `System.ServiceModel.Channels.BodyWriter` \[…\] The write state should be checked

I suspect this part of the migration advice is not necessary and Base64 implementations of `BodyWriter.OnWriteBodyContents` can unconditionally write the `Binary` element. I intend to verify that before 2019-09-02.

> When deriving from the class `System.ServiceModel.Channels.StreamBodyWriter` \[…\]

This part is OK.

**Comment by KalleOlaviNiemitalo** [comment]
@mconnew, I don't currently have a test app that I could share, and I'm not sure I will create one.

Earlier, I tested that the `writer.WriteString("")` workaround fixes the problem, when targeting .NET Framework 4.5.2 and running on .NET Framework 4.8. This was server-side WCF with `WebHttpBinding`, `transferMode="Buffered"`, `WebInvokeAttribute.ResponseFormat` = `WebMessageFormat.Json`, and an `IErrorHandler.ProvideFault` implementation that added `new WebBodyFormatMessageProperty(WebContentFormat.Raw)` to the fault `Message`. This scenario was OK with unconditionally writing the wrapper element.

I am looking at Reference Source. `System.ServiceModel.Channels.StreamBodyWriter` overrides `OnCreateBufferedCopy(int maxBufferSize)`. If an app derives a class directly from `System.ServiceModel.Channels.BodyWriter` and does not override `OnCreateBufferedCopy`, then the state will be `WriterState.Element` when `BodyWriter.OnCreateBufferedCopy` calls `OnWriteBodyContents`. If the app's implementation of `OnWriteBodyContents` was written the way you advised, it then omits the wrapper element because the state is not `WriteState.Start`, so there will be no wrapper element in the `XmlBuffer`. Later, `BufferedBodyWriter.OnWriteBodyContents` will copy the data from the `XmlBuffer` to the other `XmlDictionaryWriter`; again without the wrapper element, even if the write state was `WriteState.Start` now. I think this is a scenario that shows `OnWriteBodyContents` really needs to unconditionally write the wrapper. However, I did not test it yet.
